src/main/scala/pl/project13/kaze/KazeClass.scala

```diff
@@ -29,54 +31,117 @@ object KazeClass {
       classOf[immutable.Seq[_]].getName, classOf[Map[_, _]].getName)
 
     private val fields = clazz.getDeclaredFields
-        .filterNot(skip)
-        .filterNot(_.getName.contains("$"))
+      .filterNot(skip)
+      .filterNot(_.getName.contains("$"))
 
     def render = {
       val sb = new StringBuilder(s"${indent}final class $clazzName private(\n")
       indent = "  "
 
-      // constructor
+      renderConstructor(sb)
+
+      sb.append("{\n\n")
+
+      renderJavaAccessors(sb)
+
+      sb.append("\n")
+
+      renderWithXyz(sb)
+      sb.append("\n")
+      renderCopy(sb)
+      sb.append("\n")
+      renderToString(sb)
+      sb.append("\n")
+      renderEquals(sb)
+      sb.append("\n")
+      renderHashCode(sb)
+
+      sb.append("}\n") // end class
+
+      // companion object
+      indent = "  "
+      sb.append(s"object $clazzName {\n")
+      sb.append(s"${indent}/** Scala API */\n")
+      sb.append(s"${indent}def apply(): $clazzName = new $clazzName()\n")
+
+      sb.append(s"${indent}/** Java API */\n")
+      sb.append(s"${indent}def getInstance(): $clazzName = apply()\n")
+
+      renderObjectApply(sb)
+      renderObjectCreate(sb)
 
+      sb.append("}\n")
+
+      sb.result()
+    }
+
+    private def renderConstructor(sb: StringBuilder): Unit = {
+      val skipExtends = Set("java.lang.Object", "scala.Product", "scala.Serializable")
       fields.foreach { m =>
         val mName = m.getName
         val mType = theType(m)
         sb.append(s"${indent}val $mName: $mType,\n")
       }
-      sb.delete(sb.length - 2 , sb.length)
-      sb.append(") {\n")
-      sb.append("\n")
+      sb.delete(sb.length - 2, sb.length)
+      sb.append(")")
+      val extending = (Seq(clazz.getSuperclass.getName) ++ clazz.getInterfaces.map(_.getName))
+        .filterNot(skipExtends.contains)
+      if (extending.nonEmpty) sb.append(extending.mkString("\nextends ", "\nwith ", " "))
+    }
 
-      // with...
+    private def renderJavaAccessors(sb: StringBuilder): Unit = {
+      fields.foreach { m =>
+        sb.append(s"${indent}/** Java API */\n")
+        val mName = m.getName
+        val mType = theType(m)
+        if (mType.startsWith("Option[")) {
+          val mTypeWithoutOpt = mType.replace("Option", "java.util.Optional")
+          sb.append(s"${indent}def get${up(mName)}: $mTypeWithoutOpt = $mName.asJava\n")
+        } else if (mType.startsWith("List[")) {
```

Comment:
`immutable.Seq` should probably more common api type in Scala

Reply:
True, please amend

Reply:
I just kept List which was used before, I'll PR to change to `immutable.Seq`.

Reply:
Yeah, that was my bad then :)